File: 7-matrices/4-matrix-traversal-and-search/main.cpp
// Matrix Traversal and Search - C++ Example

#include <iostream>
#include <vector>

std::pair<int,int> searchInMatrix(const std::vector<std::vector<int>>& matrix, int target) {
    int rows = matrix.size();
    int cols = matrix[0].size();
    int r = 0, c = cols - 1;

    while (r < rows && c >= 0) {
        if (matrix[r][c] == target)
            return {r, c};
        else if (matrix[r][c] > target)
            c--;
        else
            r++;
    }
    return {-1, -1};
}

int main() {
    std::vector<std::vector<int>> matrix{
        {1,4,7,11}, {2,5,8,12}, {3,6,9,16}, {10,13,14,17}
    };

    auto pos = searchInMatrix(matrix, 5);
    std::cout << "Target 5 at: (" << pos.first << ", " << pos.second << ")" << std::endl;
    return 0;
}

File: 7-matrices/4-matrix-traversal-and-search/main.py
# Matrix Traversal and Search - Python Example

def search_in_matrix(matrix, target):
    rows, cols = len(matrix), len(matrix)
    r, c = 0, cols - 1
    while r < rows and c >= 0:
        if matrix[r][c] == target:
            return (r, c)
        elif matrix[r][c] > target:
            c -= 1
        else:
            r += 1
    return (-1, -1)

matrix = [
    [1, 4, 7, 11],
    [2, 5, 8, 12],
    [3, 6, 9, 16],
    [10,13,14,17]
]

print("Target 5 at:", search_in_matrix(matrix, 5))
